lib/pure/times.nim

```diff
@@ -1263,14 +1242,28 @@ when not defined(JS):
     # copying is needed anyway to provide reentrancity; thus
     # the conversion is not expensive
 
-  proc toTime(timeInfo: TimeInfo): Time =
-    var cTimeInfo = timeInfo # for C++ we have to make a copy
-    # because the header of mktime is broken in my version of libc
+  proc toEpochday(year, month, day: int): int64 =
+    # Based on http://howardhinnant.github.io/date_algorithms.html
+    var (y, m, d) = (year, month, day)
+    if m <= 2:
+      y.dec
+
+    let era = (if y >= 0: y else: y-399) div 400;
+    let yoe = y - era * 400
+    let doy = (153 * (m + (if m > 2: -3 else: 9)) + 2) div 5 + d-1; 
+    let doe = yoe * 365 + yoe div 4 - yoe div 100 + doy;         
+    return era * 146097 + doe - 719468;
 
-    result = mktime(timeInfoToTM(cTimeInfo))
-    # mktime is defined to interpret the input as local time. As timeInfoToTM
-    # does ignore the timezone, we need to adjust this here.
-    result = Time(TimeImpl(result) - getTimezone() + timeInfo.timezone)
+  proc toTime(timeInfo: TimeInfo): Time =
+    let epochDay = toEpochday(timeInfo.year, ord(timeInfo.month) + 1, timeInfo.monthday)
+    result = Time(epochDay * secondsInDay)
+    result.inc timeInfo.hour * secondsInHour
+    result.inc timeInfo.minute * 60
+    result.inc timeInfo.second
+    # The code above treats `ti` as GM,
+    # so we need to compensate for that here
```

Comment:
Thank you @GULPF for your work on the times module! 
I think the comment is not correct. Its exactly the other way round: the code treats ti as local and you adjust it to UTC.
Maybe it would be useful to also add a pure Nim fromTime procedure. If you want to see, how i implemented it for my own DateTime module, please see http://github.com/skilchen/DateTime

Reply:
Hmm I think the comment is correct. The first part ignores the UTC offset and generates a timestamp "as if" the offset was 0. The second part adjusts the timestamp so that it takes the original `ti`'s offset into account.

The issue with `fromTime` in pure Nim is that it's impossible to generate a local `TimeInfo` without having full information about the local timezone, which we don't. By using the C stdlib (which do have full tz information) we get that functionality for free :)

Reply:
It seems that the comment could be correct. But that's an illusion. The stringification of the time value - unfortunately - calls getLocalTime on the passed time value.

The code you ported from H. Hinnant ignores the timezone competely. It simply takes what is stored in the relevant TimeInfo fields. Then you convert it to UTC and the stringification converts it back to local time ... 

The logic you apply is indeed the logic to convert from local to UTC time. The timezone offset is stored in seconds west of UTC. So if you are - lets say - 8 hours west of UTC then you have the value 28800 in timezone. If you add this to the stored time, you move the time in the eastward direction of Grenwich. So if it is 12:00 in your location it is 20:00 in Grenwich, which is what you get adding 8 hours to your local time ...

Reply:
I'm not sure I understand. I know that `toEpochDay` ignores timezone - that's the same as treating it as UTC. So after the first part, `result` is a timestamp representing `ti` if we pretend that `ti.timezone = 0` and `ti.isDst = false`. That is, if we pretend that `ti` is already in UTC. The second part corrects the timestamp so it takes the timezone/dst into account. It doesn't convert from local to UTC, because `result` doesn't have a timezone.

Either way, the comment is not great. What about this:  
```nim
# The code above ignores the UTC offset of `timeInfo`,
# so we need to compensate for that here`
```

Reply:
What you usually have in a timeInfo is a local date/time and and an offset of that time to UTC. Your toTime procedure returns the time in UTC. I would prefer the following comment:
```
  # we return the time in UTC
```
If you want to get the same timeInfo back, you have to call:
```nim
getLocalTime(toTime(ti))
```
My understanding is that getLocalTime converts a time expressed in UTC to your local time zone.
Maybe a little example helps to clarify things: 
https://play.nim-lang.org/?gist=73b460d5345c850d409795fe7d680df5

Your toTime behaves exactly the same way as the old one. So it doesn't matter that your procedure is not yet available in the playground. Please pay special attention to the lines beginning with "diff to UTC". It should be 0 in all three cases, regardless of the current timezone settings, otherwise my assumptions are all wrong.

(Btw. The comment in the original is also misleading. Because the line
```nim
Time(TimeImpl(result) - getTimezone() + timeInfo.timezone)
```
doesn't adjust anything. If you are in CET then getTimezone() returns -3600 and in timeInfo.timezone you hopefully also have -3600. So it first subtracts value x and then adds the same value x back. If you manually changed timeInfo.timezone to something else, then the adjustment made here gives no reasonable result.

Reply:
Neither `getTime` or `toTime` returns the time in UTC. They return Unix timestamps, which Wikipedia defines like this: "defined as the number of seconds that have elapsed since 00:00:00 Coordinated Universal Time (UTC), Thursday, 1 January 1970". By that definition, a Unix timestamp can't have a timezone. I guess that theoretically you could have a Unix timestamp that represents the number of second since 1970-01-01T00:00:00 local time, but that would be useless in practice since the whole point of a timestamp is to unambiguously represent some point in time.

>(Btw. The comment in the original is also misleading. Because the line
>
>```Time(TimeImpl(result) - getTimezone() + timeInfo.timezone)```
>
>doesn't adjust anything. If you are in CET then getTimezone() returns -3600 and in timeInfo.timezone you hopefully also have -3600. So it first subtracts value x and then adds the same value x back. If you manually changed timeInfo.timezone to something else, then the adjustment made here gives no reasonable result.

It does adjust something, although incorrectly (because `getTimezone()` might not return the offset that was active for the point in time that `timeInfo` represents - sometimes timezones change their offset). The times module is supposed to support any timezone, it just doesn't do it properly at the moment.

Reply:
So... is this dispute settled? Can we merge it? I'm not an expert on times so I'll trust GULPF's implementation.

Reply:
Well the implementation is correct, just some bikeshedding about the comment :P I've now changed the comment to what I suggested earlier, which imo is good enough.

Reply:
As I've mentioned in another PR for times, let's implement a testing foundation for the ``times`` module. And let's do it before merging any of these PRs.